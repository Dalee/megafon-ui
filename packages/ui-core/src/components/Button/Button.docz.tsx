import * as React from 'react';

export const BlockWrapper = ({ children }) => {
    return (
        <div style={{
            padding: '20px',
            margin: '-20px',
            backgroundColor: '#F5F5F5',
        }}>
            {children}
        </div>
    );
};

export const DarkBlockWrapper = ({ children }) => {
    return (
        <div style={{
            padding: '20px',
            margin: '-20px',
            backgroundColor: '#333333',
        }}>
            {children}
        </div>
    );
};

export const ButtonWrapper = ({ children }) => {
    return (
        <div style={{
            display: 'inline-block',
            marginRight: '25px',
        }}>
            {children}
        </div>
    );
};
