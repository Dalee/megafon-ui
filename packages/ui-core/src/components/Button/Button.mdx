---
name: Button
menu: Components
route: components/button
---

import { Playground, Props } from 'docz';
import Button from './Button';
import Balance from 'icons/Basic/24/Balance_24.svg';
import { BlockWrapper, ButtonWrapper } from './Button.docz';

# Button

## Properties

<Props of={Button} />

## Basic usage: type = primary, theme = green

<Playground>
    <BlockWrapper>
        <ButtonWrapper>
            <Button>active</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button disabled>disabled</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button showSpinner>with spinner</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button showArrow>with arrow</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button iconLeft=<Balance />>with icon</Button>
        </ButtonWrapper>
    </BlockWrapper>
</Playground>

## Basic usage: type = primary, theme = purple

<Playground>
    <BlockWrapper>
        <ButtonWrapper>
            <Button theme="purple">active</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button theme="purple" disabled>disabled</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button theme="purple" showSpinner>with spinner</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button theme="purple" showArrow>with arrow</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button theme="purple" iconLeft=<Balance />>with icon</Button>
        </ButtonWrapper>
    </BlockWrapper>
</Playground>

## Basic usage: type = primary, theme = white

<Playground>
    <BlockWrapper>
        <ButtonWrapper>
            <Button theme="white">active</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button theme="white" disabled>disabled</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button theme="white" showSpinner>with spinner</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button theme="white" showArrow>with arrow</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button theme="white" iconLeft=<Balance />>with icon</Button>
        </ButtonWrapper>
    </BlockWrapper>
</Playground>


## Basic usage: type = outline, theme = green

<Playground>
    <BlockWrapper>
        <ButtonWrapper>
            <Button type="outline">active</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" disabled>disabled</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" showSpinner>with spinner</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" showArrow>with arrow</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" iconLeft=<Balance />>with icon</Button>
        </ButtonWrapper>
    </BlockWrapper>
</Playground>

## Basic usage: type = outline, theme = purple

<Playground>
    <BlockWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="purple">active</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="purple" disabled>disabled</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="purple" showSpinner>with spinner</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="purple" showArrow>with arrow</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="purple" iconLeft=<Balance />>with icon</Button>
        </ButtonWrapper>
    </BlockWrapper>
</Playground>

## Basic usage: type = outline, theme = white

<Playground>
    <BlockWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="white">active</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="white" disabled>disabled</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="white" showSpinner>with spinner</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="white" showArrow>with arrow</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="white" iconLeft=<Balance />>with icon</Button>
        </ButtonWrapper>
    </BlockWrapper>
</Playground>

## Basic usage: type = outline, theme = black

<Playground>
    <BlockWrapper customStyles={{
        backgroundColor: '#333333'
    }}>
        <ButtonWrapper>
            <Button type="outline" theme="black">active</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="black" disabled>disabled</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="black" showSpinner>with spinner</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="black" showArrow>with arrow</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button type="outline" theme="black" iconLeft=<Balance />>with icon</Button>
        </ButtonWrapper>
    </BlockWrapper>
</Playground>

## Button size

<Playground>
    <BlockWrapper>
        <ButtonWrapper>
            <Button sizeAll="small" showArrow>Small</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button sizeAll="medium" iconLeft=<Balance />>Medium</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button sizeAll="large">Large</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button sizeDesktop="large" sizeTablet="medium" sizeMobile="small" showArrow>Button</Button>
        </ButtonWrapper>
    </BlockWrapper>
</Playground>

## Full width button

<Playground>
    <BlockWrapper>
        <ButtonWrapper customStyles={{ width: '100%' }}>
            <Button fullWidth>Full Width Button</Button>
        </ButtonWrapper>
    </BlockWrapper>
</Playground>

## Link button

<Playground>
    <BlockWrapper>
        <ButtonWrapper>
            <Button href="#">Link</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button href="#" disabled>Disabled Link</Button>
        </ButtonWrapper>
        <ButtonWrapper>
            <Button href="#" target="_blank">Target Link</Button>
        </ButtonWrapper>
    </BlockWrapper>
</Playground>
