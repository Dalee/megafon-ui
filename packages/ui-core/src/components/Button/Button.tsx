import * as React from 'react';
import PropTypes from 'prop-types';
import cnCreate from 'utils/cnCreate';
import './Button.less';
import Spinner from 'docIcons/spinner.svg';
import Arrow from 'icons/System/32/Arrow_right_32.svg';
import detectTouch from 'utils/detectTouch';
import filterDataAttrs, { IDataAttributes } from './../../utils/dataAttrs';

export interface IButtonProps extends IDataAttributes {
    /** Дополнительный класс корневого элемента */
    className?: string | string[];
    /** Дополнительные классы для внутренних элементов */
    classes?: {
        /** Button class */
        root?: string;
        /** Content class */
        content?: string;
        /** Inner container class */
        inner?: string;
    };
    /** Тема компонента */
    theme?: 'green' | 'purple' | 'white' | 'black';
    /** Тип компонента */
    type?: 'primary' | 'outline';
    /** Ссылка */
    href?: string;
    /** Target - свойство тега <a> */
    target?: '_self' | '_blank' | '_parent' | '_top';
    /** Поведение кнопки */
    actionType?: 'button' | 'reset' | 'submit';
    /** Размер на всех разрешениях экрана */
    sizeAll?: 'small' | 'medium' | 'large';
    /** Размер на разрешении экрана 1280+ */
    sizeWide?: 'small' | 'medium' | 'large';
    /** Размер на разрешении экрана 1020+ */
    sizeDesktop?: 'small' | 'medium' | 'large';
    /** Размер на разрешении экрана 730-1020 */
    sizeTablet?: 'small' | 'medium' | 'large';
    /** Размер на разрешении экрана 320-730 */
    sizeMobile?: 'small' | 'medium' | 'large';
    /** Растянуть на полную ширину контейнера */
    fullWidth?: boolean;
    /** Показать загрузку */
    showSpinner?: boolean;
    /** Показать стелку */
    showArrow?: boolean;
    /** Иконка слева */
    iconLeft?: JSX.Element;
    /** Управление возможности взаимодействия с компонентом */
    disabled?: boolean;
    /** Обработчик клика по кнопке */
    onClick?: (e: React.SyntheticEvent<EventTarget>) => void;
}

const cn = cnCreate('mfui-beta-button');
const Button: React.FC<IButtonProps> = props => {
    const {
        classes: {
            root: rootClassName,
            content: contentClassName,
            inner: innerClassName,
        } = {},
        className = '',
        theme = 'green',
        type = 'primary',
        href,
        target,
        actionType = 'button',
        sizeAll = 'medium',
        sizeWide,
        sizeDesktop,
        sizeTablet,
        sizeMobile,
        fullWidth = false,
        showSpinner = false,
        showArrow = false,
        iconLeft,
        disabled,
        children,
        onClick,
        dataAttrs,
    } = props;

    const isTouch: boolean = React.useMemo(() => detectTouch(), []);
    const ElementType = href ? 'a' : 'button';

    const handleClick = React.useCallback((e: React.SyntheticEvent<EventTarget>): void => {
        if (disabled) {
            e.preventDefault();

            return;
        }

        onClick && onClick(e);
    }, [disabled, onClick]);

    const currentTheme: string = React.useMemo(() => (
        (type === 'primary') && (theme === 'black') ? 'green' : theme
    ), [theme]);

    const renderChildren: JSX.Element = React.useMemo(() => (
        <div className={cn('content', contentClassName)}>
            {iconLeft && <div className={cn('icon')}>{iconLeft}</div>}
            {children}
            {!iconLeft && showArrow && <Arrow className={cn('icon-arrow')} />}
        </div>
    ), [contentClassName, showArrow, children]);

    const renderSpinner: JSX.Element = React.useMemo(() => (
        <div className={cn('spinner')}>
            <Spinner />
        </div>
    ), []);

    return (
        <ElementType
            {...filterDataAttrs(dataAttrs)}
            className={cn({
                type,
                theme: currentTheme,
                disabled,
                'size-all': sizeAll,
                'size-wide': sizeWide,
                'size-desktop': sizeDesktop,
                'size-tablet': sizeTablet,
                'size-mobile': sizeMobile,
                'full-width': fullWidth,
                loading: showSpinner,
                'no-touch': !isTouch,
            }, [className, rootClassName])}
            href={href}
            target={href ? target : undefined}
            type={href ? undefined : actionType}
            onClick={handleClick}
            disabled={!href && disabled}
        >
            <div className={cn('inner', innerClassName)}>
                {!showSpinner && children && renderChildren}
                {showSpinner && renderSpinner}
            </div>
        </ElementType>
    );
};

Button.propTypes = {
    classes: PropTypes.shape({
        root: PropTypes.string,
        content: PropTypes.string,
        inner: PropTypes.string,
    }),
    theme: PropTypes.oneOf(['green', 'purple', 'white', 'black']),
    type: PropTypes.oneOf(['primary', 'outline']),
    href: PropTypes.string,
    target: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']),
    actionType: PropTypes.oneOf(['button', 'reset', 'submit']),
    sizeAll: PropTypes.oneOf(['small', 'medium', 'large']),
    sizeWide: PropTypes.oneOf(['small', 'medium', 'large']),
    sizeDesktop: PropTypes.oneOf(['small', 'medium', 'large']),
    sizeTablet: PropTypes.oneOf(['small', 'medium', 'large']),
    sizeMobile: PropTypes.oneOf(['small', 'medium', 'large']),
    fullWidth: PropTypes.bool,
    showSpinner: PropTypes.bool,
    showArrow: PropTypes.bool,
    iconLeft: PropTypes.element,
    disabled: PropTypes.bool,
    dataAttrs: PropTypes.objectOf(PropTypes.string.isRequired),
    onClick: PropTypes.func,
};

export default Button;
