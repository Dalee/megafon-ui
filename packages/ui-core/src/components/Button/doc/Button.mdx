---
name: Button
menu: Компоненты
route: components/button
---

import { Props } from 'docz';
import Button from '../Button';
import Example from './Button.example.mdx';

# Button

<ImportExample name="Button" />

<DoczTabs>
    <Example />
    <Props of={Button} />
</DoczTabs>
