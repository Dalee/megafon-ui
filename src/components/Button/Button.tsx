import * as React from 'react';
import * as PropTypes from 'prop-types';
import { cnCreate } from '../../utils/cn';
import './Button.less';
import Spinner from 'icons/spinner.svg';

interface Props {
    /** Special view */
    customView?: 'two-lines';
    /** Link */
    href?: string | null;
    /** Target - property tag <a> */
    target?: '_self' | '_blank' | '_parent' | '_top';
    /** Functional for form */
    type?: 'button' | 'reset' | 'submit';
    /** Size for all devices */
    sizeAll?: 'small' | 'medium' | 'large';
    /** Size for wide devices 1280+ */
    sizeWide?: 'small' | 'medium' | 'large';
    /** Size for desktop 1020+ */
    sizeDesktop?: 'small' | 'medium' | 'large';
    /** Size for tablet 730-1020 */
    sizeTablet?: 'small' | 'medium' | 'large';
    /** Size for mobile 320-730 */
    sizeMobile?: 'small' | 'medium' | 'large';
    /** Current color */
    passiveColor?: 'green' | 'purple' | 'transparent' | 'transparent-green' | 'white';
    /** Hover color */
    hoverColor?: 'green' | 'purple' | 'transparent';
    /** Click/press color */
    downColor?: 'dark' | 'transparent';
    /** Disabled color */
    disabledColor?: 'gray' | 'white' | 'transparent';
    /** Width */
    width?: 'full' | 'auto';
    /** Margin(outer indentation) */
    margin?: boolean;
    /** Disabled */
    disabled?: boolean;
    /** Padding(inner indentation) */
    padding?: boolean;
    /** Show spinner */
    showSpinner?: boolean;
    children?: JSX.Element[] | Element[] | JSX.Element | string | Element;
    /** Click event handler */
    onClick?(e: React.SyntheticEvent<EventTarget>): void;
}

const cn = cnCreate('button');
class Button extends React.Component<Props, {}> {
    static propTypes = {
        customView: PropTypes.oneOf(['two-lines']),
        href: PropTypes.string,
        target: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']),
        type: PropTypes.oneOf(['button', 'reset', 'submit']),
        disabled: PropTypes.bool,
        sizeAll: PropTypes.oneOf(['small', 'medium', 'large']),
        sizeWide: PropTypes.oneOf(['small', 'medium', 'large']),
        sizeDesktop: PropTypes.oneOf(['small', 'medium', 'large']),
        sizeTablet: PropTypes.oneOf(['small', 'medium', 'large']),
        sizeMobile: PropTypes.oneOf(['small', 'medium', 'large']),
        passiveColor: PropTypes.oneOf([
            'green',
            'purple',
            'transparent',
            'transparent-green',
            'white',
        ]),
        hoverColor: PropTypes.oneOf([
            'green',
            'purple',
            'transparent',
        ]),
        downColor: PropTypes.oneOf([
            'dark',
            'transparent',
        ]),
        disabledColor: PropTypes.oneOf([
            'gray',
            'white',
            'transparent',
        ]),
        margin: PropTypes.bool,
        padding: PropTypes.bool,
        width: PropTypes.oneOf(['full', 'auto']),
        showSpinner: PropTypes.bool,
        onClick: PropTypes.func,
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
            PropTypes.node,
        ]),
    };

    static defaultProps = {
        passiveColor: 'green',
        hoverColor: 'green',
        downColor: 'dark',
        disabledColor: 'gray',
        width: 'auto',
        type: 'button',
        sizeAll: 'medium',
        padding: true,
        showSpinner: false,
    };

    renderChildrenElem() {
        return (
            <div className={cn('content')}>
                {this.props.children}
            </div>
        );
    }

    renderSpinner() {
        return (
            <div className={cn('spinner')}>
                <Spinner />
            </div>
        );
    }

    render() {
        const ElementType = this.props.href ? 'a' : 'button';

        return (
            <ElementType
                className={cn('', {
                    'size-all': this.props.sizeAll,
                    'size-wide': this.props.sizeWide,
                    'size-desktop': this.props.sizeDesktop,
                    'size-tablet': this.props.sizeTablet,
                    'size-mobile': this.props.sizeMobile,
                    'custom-view': this.props.customView,
                    'passive-color': !this.props.customView && this.props.passiveColor,
                    'hover-color': !this.props.customView && this.props.hoverColor,
                    'down-color': !this.props.customView && this.props.downColor,
                    'disabled-color': !this.props.customView && this.props.disabledColor,
                    padding: this.props.padding,
                    width: this.props.width,
                    margin: this.props.margin,
                    loading: this.props.showSpinner,
                })}
                href={this.props.href}
                target={this.props.target}
                type={this.props.href ? '' : this.props.type}
                onClick={this.props.onClick}
                disabled={this.props.disabled}>
                <div className={cn('inner')}>
                    {!this.props.showSpinner && this.props.children && this.renderChildrenElem()}
                    {this.props.showSpinner && this.renderSpinner()}
                </div>
            </ElementType>
        );
    }
}

export default Button;
