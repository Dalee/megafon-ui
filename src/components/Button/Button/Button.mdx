---
name: Button
menu: Components
---

import { Playground, PropsTable } from 'docz';
import Button from './Button';
import ArrowBack from 'icons/arrow-back.svg';

# Button

## Properties

<PropsTable of={Button} />

## Basic usage

<Playground>
    <Button>Кнопка</Button>
</Playground>

## Spinner usage

<Playground>
    <Button showSpinner={true}>Кнопка</Button>
</Playground>
