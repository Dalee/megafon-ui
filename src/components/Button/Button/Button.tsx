import * as React from 'react';
import * as PropTypes from 'prop-types';
import { cnCreate } from 'utils/cn';
import './Button.less';
import Spinner from 'icons/spinner.svg';

interface Props {
    /** Spicial view */
    customView?: 'arrow-back' | 'two-lines';
    /** Link */
    href?: string | null;
    /** target - property tag <a> */
    target?: '_self' | '_blank' | '_parent' | '_top';
    /** Functional */
    type?: 'button' | 'reset' | 'submit';
    /** Size for all devices */
    sizeAll?: 'small' | 'medium' | 'large';
    /** Size for wide devices */
    sizeWide?: 'small' | 'medium' | 'large';
    /** Size for desktop */
    sizeDesktop?: 'small' | 'medium' | 'large';
    /** Size for tablet */
    sizeTablet?: 'small' | 'medium' | 'large';
    /** Size for mobile */
    sizeMobile?: 'small' | 'medium' | 'large';
    /** Current color */
    passiveColor?: 'green' | 'purple' | 'transparent' | 'transparent-green' | 'white';
    /** Hover color */
    hoverColor?: 'green' | 'purple' | 'transparent';
    /** Click/press color */
    downColor?: 'dark' | 'transparent';
    /** Disabled color */
    disabledColor?: 'gray' | 'white' | 'transparent';
    /** Width */
    width?: 'full' | 'auto';
    /** margin(outer indentation) */
    margin?: boolean;
    /** Disabled */
    disabled?: boolean;
    /** Padding(inner indentation) */
    padding?: boolean;
    /** Show spinner */
    showSpinner?: boolean;
    children?: JSX.Element[] | Element[] | JSX.Element | string | Element;
    /** Svg icon */
    svgIcon?: JSX.Element;
    /** Click event handler */
    onClick?(e: React.SyntheticEvent<EventTarget>): void;
}

/**
 * Компонент кнопки
 */
const cn = cnCreate('button');
class Button extends React.Component<Props, {}> {
    static propTypes = {
        /** Кастомный внешний вид */
        customView: PropTypes.oneOf(['arrow-back', 'two-lines']),
        /** Адрес ссылки */
        href: PropTypes.string,
        target: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']),
        /** Тип */
        type: PropTypes.oneOf(['button', 'reset', 'submit']),
        /** Заблокировать кнопку */
        disabled: PropTypes.bool,
        /** Размер кнопки на всех разрешениях */
        sizeAll: PropTypes.oneOf(['small', 'medium', 'large']),
        /** Размер кнопки на разрешении 1280+ */
        sizeWide: PropTypes.oneOf(['small', 'medium', 'large']),
        /** Размер кнопки на разрешении 1020+ */
        sizeDesktop: PropTypes.oneOf(['small', 'medium', 'large']),
        /** Размер кнопки на разрешении 730-1020 */
        sizeTablet: PropTypes.oneOf(['small', 'medium', 'large']),
        /** Размер кнопки на разрешении 320-730 */
        sizeMobile: PropTypes.oneOf(['small', 'medium', 'large']),
        /** Цвет кнопки в пасивном состоянии */
        passiveColor: PropTypes.oneOf([
            'green',
            'purple',
            'transparent',
            'transparent-green',
            'white',
        ]),
        /** Цвет кнопки в наведенном состоянии */
        hoverColor: PropTypes.oneOf([
            'green',
            'purple',
            'transparent',
        ]),
        /** Цвет кнопки в нажатом состоянии */
        downColor: PropTypes.oneOf([
            'dark',
            'transparent',
        ]),
        /** Цвет кнопки в отключенном состоянии */
        disabledColor: PropTypes.oneOf([
            'gray',
            'white',
            'transparent',
        ]),
        /** Использовать внешний правый и нижний отступы */
        margin: PropTypes.bool,
        /** Использовать внутренний правый и левый отступы */
        padding: PropTypes.bool,
        /** Растянуть на всю ширину */
        width: PropTypes.oneOf(['full', 'auto']),
        /** Показать спиннер */
        showSpinner: PropTypes.bool,
        /**
         * Обработчик события onClick
         *
         * @param {SyntheticEvent} event - синтетическое React событие.
         */
        onClick: PropTypes.func,
        /** Svgicon */
        svgIcon: PropTypes.element,
        /** Дочерние элементы */
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
            PropTypes.node,
        ]),
    };

    static defaultProps = {
        passiveColor: 'green',
        hoverColor: 'green',
        downColor: 'dark',
        disabledColor: 'gray',
        width: 'auto',
        type: 'button',
        sizeAll: 'medium',
        padding: true,
        showSpinner: false,
    };

    renderSvgIcon() {
        return (
            <div className={cn('icon-box')}>
                <div className={cn('icon')}>
                    {this.props.svgIcon}
                </div>
            </div>
        );
    }

    renderChildrenElem() {
        return (
            <div className={cn('content', { icon: !!this.props.svgIcon })}>
                {this.props.children}
            </div>
        );
    }

    renderSpinner() {
        return (
            <div className={cn('spinner')}>
                <Spinner />
            </div>
        );
    }

    render() {
        const ElementType = this.props.href ? 'a' : 'button';

        return (
            <ElementType
                className={cn('', {
                    'size-all': this.props.sizeAll,
                    'size-wide': this.props.sizeWide,
                    'size-desktop': this.props.sizeDesktop,
                    'size-tablet': this.props.sizeTablet,
                    'size-mobile': this.props.sizeMobile,
                    'custom-view': this.props.customView,
                    'passive-color': !this.props.customView && this.props.passiveColor,
                    'hover-color': !this.props.customView && this.props.hoverColor,
                    'down-color': !this.props.customView && this.props.downColor,
                    'disabled-color': !this.props.customView && this.props.disabledColor,
                    padding: this.props.padding,
                    width: this.props.width,
                    margin: this.props.margin,
                    loading: this.props.showSpinner,
                })}
                href={this.props.href}
                target={this.props.target}
                type={this.props.href ? '' : this.props.type}
                onClick={this.props.onClick}
                disabled={this.props.disabled}>
                <div className={cn('inner')}>
                    {!this.props.showSpinner && this.props.svgIcon && this.renderSvgIcon()}
                    {!this.props.showSpinner && this.props.children && this.renderChildrenElem()}
                    {this.props.showSpinner && this.renderSpinner()}
                </div>
            </ElementType>
        );
    }
}

export default Button;
