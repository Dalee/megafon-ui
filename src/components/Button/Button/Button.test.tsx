import * as React from 'react';
import { shallow, mount } from 'enzyme';
import Button from './Button';
import ArrowBack from 'icons/arrow-back.svg';

describe('<Button />', () => {
    it('it renders Button', () => {
        const wrapper = shallow(<Button />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders children', () => {
        const wrapper = shallow(<Button>button</Button>);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders icon', () => {
        const wrapper = mount(<Button svgIcon={<ArrowBack />} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders spinner', () => {
        const wrapper = mount(<Button showSpinner={true} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders tag a', () => {
        const wrapper = mount(<Button href="http://moscow.megafon.ru/" />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it calls click handler', () => {
        const onClick = jest.fn();
        const wrapper = mount(<Button href="http://moscow.megafon.ru/" onClick={onClick} />);
        wrapper.simulate('click');
        expect(onClick).toHaveBeenCalledTimes(1);
    });
});
