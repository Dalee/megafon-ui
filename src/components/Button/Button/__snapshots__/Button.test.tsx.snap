// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> it renders Button 1`] = `
<button
  className="button button_size-all_medium button_passive-color_green button_hover-color_green button_down-color_dark button_disabled-color_gray button_padding button_width_auto"
  type="button"
>
  <div
    className="button__inner"
  />
</button>
`;

exports[`<Button /> it renders children 1`] = `
<button
  className="button button_size-all_medium button_passive-color_green button_hover-color_green button_down-color_dark button_disabled-color_gray button_padding button_width_auto"
  type="button"
>
  <div
    className="button__inner"
  >
    <div
      className="button__content"
    >
      button
    </div>
  </div>
</button>
`;

exports[`<Button /> it renders icon 1`] = `
<Button
  disabledColor="gray"
  downColor="dark"
  hasArrowBack={true}
  hoverColor="green"
  padding={true}
  passiveColor="green"
  showSpinner={false}
  sizeAll="medium"
  type="button"
  width="auto"
>
  <button
    className="button button_size-all_medium button_passive-color_green button_hover-color_green button_down-color_dark button_disabled-color_gray button_padding button_width_auto"
    type="button"
  >
    <div
      className="button__inner"
    >
      <div
        className="button__icon-box"
      >
        <test-file-stub
          className="button__icon"
        />
      </div>
    </div>
  </button>
</Button>
`;

exports[`<Button /> it renders spinner 1`] = `
<Button
  disabledColor="gray"
  downColor="dark"
  hoverColor="green"
  padding={true}
  passiveColor="green"
  showSpinner={true}
  sizeAll="medium"
  type="button"
  width="auto"
>
  <button
    className="button button_size-all_medium button_passive-color_green button_hover-color_green button_down-color_dark button_disabled-color_gray button_padding button_width_auto button_loading"
    type="button"
  >
    <div
      className="button__inner"
    >
      <div
        className="button__spinner"
      >
        <test-file-stub />
      </div>
    </div>
  </button>
</Button>
`;

exports[`<Button /> it renders tag a 1`] = `
<Button
  disabledColor="gray"
  downColor="dark"
  hoverColor="green"
  href="http://moscow.megafon.ru/"
  padding={true}
  passiveColor="green"
  showSpinner={false}
  sizeAll="medium"
  type="button"
  width="auto"
>
  <a
    className="button button_size-all_medium button_passive-color_green button_hover-color_green button_down-color_dark button_disabled-color_gray button_padding button_width_auto"
    href="http://moscow.megafon.ru/"
    type=""
  >
    <div
      className="button__inner"
    />
  </a>
</Button>
`;
